### PR TITLE
Add Iterable.slice() and Iterable.subsequence() extensions

### PR DESCRIPTION
**Pull request by nex3**
This also adds more efficient implementations of these for List.

Inspired by Ruby's Enumerable.each_slice and Enumerable.each_cons
methods.

### CODE REVIEW

File: lib/src/iterable_extensions.dart [hunk 1]
```diff
@@ -549,6 +550,54 @@ extension IterableExtension<T> on Iterable<T> {
...
+
+  /// Returns an iterable of each subsequence of [this] of the given [length].
+  ///
+  /// Each subsequence is exactly [length] elements long and shares `length - 1`
+  /// elements with the previous subsequence. Returns an empty iterable if
+  /// [length] is greater than this iterable's length.
+  ///
+  /// For example, `(1, 2, 3, 4).subsequences(2)` returns
+  /// `([1, 2], [2, 3], [3, 4])`.
+  Iterable<List<T>> subsequences(int length) sync* {
```
Comment (lrhn) [inline lib/src/iterable_extensions.dart:584]: The name also is not particularly descriptive. Maybe `allSubsequences`?

File: lib/src/iterable_extensions.dart [hunk 2]
```diff
@@ -549,6 +550,54 @@ extension IterableExtension<T> on Iterable<T> {
...
   }
+
+  /// Returns an iterable whose elements are contiguous slices of [this].
+  ///
+  /// Each slice is [length] elements long, except for the last one which may be
+  /// shorter if [this] contains too few elements. Each slice begins after the
+  /// last one ends.
+  ///
+  /// For example, `(1, 2, 3, 4, 5).slices(2)` returns `((1, 2), (3, 4), (5))`.
+  Iterable<Iterable<T>> slices(int length) sync* {
```
Comment (lrhn) [inline lib/src/iterable_extensions.dart:561]: The "slices" name is not particularly telling.
Not sure which short-ish verb would better describe "chop list into equal sized chunks" through.
Ideas welcome.

**Review by lrhn — CHANGES_REQUESTED** [review]

File: lib/src/iterable_extensions.dart (hunk 2 above)
Comment (nex3) [reply lib/src/iterable_extensions.dart]: Ruby calls it `each_slice`, from which I got this name. RxJS calls it `bufferWithCount`, but that seems too strongly tied to its streaming API to use here. A Haskell package I found calls it `chunksOf`.

I suppose we could call this `chunks`, but I'm not sure that's much clearer.

File: lib/src/iterable_extensions.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/iterable_extensions.dart:581]: > What is the use-case for this operation. It seems fairly inefficient.

It's useful with `length = 2` for things like creating connections between a list of elements (for example, I want to use it to connect the stdout and stdin streams of a sequence of process objects). It allows calling code to avoid manually handling the fencepost problem.

If you don't think it's worth adding, I'm not opposed to removing it. I definitely think `slices()` is the more widely applicable API, this is just the natural dual so it made sense to me to include both at once.

> Could you collect the elements into a single list, then emit views on that list. It will keep older elements alive (but so will the view used by the `List` implementation)

I think the worst-case problems with unexpectedly retaining the entire iterable contents in memory are likely to be a bigger problem than copying elements each time, especially because `length` is likely to usually be very small.

Comment (nex3) [reply lib/src/iterable_extensions.dart:581]: I'm not sure "all" adds much descriptive power there... if you wanted it to be more explicit, it could be `subsequencesOfLength`, but that's a bit redundant with the parameter name.

File: lib/src/iterable_extensions.dart [hunk 3]
```diff
@@ -549,6 +550,54 @@ extension IterableExtension<T> on Iterable<T> {
...
+  /// [length] is greater than this iterable's length.
+  ///
+  /// For example, `(1, 2, 3, 4).subsequences(2)` returns
+  /// `([1, 2], [2, 3], [3, 4])`.
+  Iterable<List<T>> subsequences(int length) sync* {
+    if (length < 1) throw RangeError.range(length, 1, null, 'length');
+
+    var iterator = this.iterator;
+    var subsequence = <T>[];
+    while (iterator.moveNext()) {
```
[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/iterable_extensions.dart]: > I'd probably split this into two loops, instead of branching in the main loop on something which is going to always be false up to some point, then always true after. It really is two loops.

Done.

> Maybe:
> 
> ```dart
> var buffer = ListQueue<T>(length + 1);
> while (buffer.length < length) {
>   if (iterator.moveNext()) {
>     buffer.add(iterator.current);
>   } else {
>     return;
>   }
> }
> while (true) {
>   yield List<T>.unmodifiable(buffer);
>   if (!iterator.moveNext()) return;
>   buffer..removeFirst()..add(iterator.current);
> }
> ```

I don't think the `ListQueue` actually makes this more efficient. You're still copying `length` elements on every iteration in the `List.unmodifiable()` call.

File: lib/src/list_extensions.dart [hunk 4]
```diff
@@ -252,6 +252,38 @@ extension ListExtensions<E> on List<E> {
     }
     return true;
   }
+
+  /// Returns an iterable whose elements are contiguous [slice]s of [this].
```
[the remark this answers is not in the document]
Comment (natebosch) [reply lib/src/list_extensions.dart:256]: "Returns" is very common.

cc @munificent for opinions on this.

File: lib/src/iterable_extensions.dart [hunk 5]
```diff
@@ -549,6 +549,54 @@ extension IterableExtension<T> on Iterable<T> {
...
+  /// For example, `{1, 2, 3, 4, 5}.slices(2)` returns `([1, 2], [3, 4], [5])`.
+  Iterable<List<T>> slices(int length) sync* {
+    if (length < 1) throw RangeError.range(length, 1, null, 'length');
+
+    var iterator = this.iterator;
+    while (iterator.moveNext()) {
+      var slice = [iterator.current];
+      for (var i = 1; i < length && iterator.moveNext(); i++) {
+        slice.add(iterator.current);
+      }
```
[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/iterable_extensions.dart:568]: I think that's less readable ultimately. `--i` is not something you encounter frequently in Dart code.

File: lib/src/iterable_extensions.dart [hunk 6]
```diff
@@ -549,6 +549,54 @@ extension IterableExtension<T> on Iterable<T> {
...
+      if (iterator.moveNext()) {
+        subsequence.add(iterator.current);
+      } else {
+        return;
+      }
+    }
+
+    yield subsequence;
+    while (iterator.moveNext()) {
+      subsequence = [...subsequence.skip(1), iterator.current];
```
[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/iterable_extensions.dart:596]: On the other hand, you could imagine `...value` inserting a runtime check for `EfficientLengthIterable`! I think the internals of the compiler are enough of a black box that it mostly makes sense to make the code as readable and semantic as possible.

**Review by nex3 — COMMENTED** [review]
Holding off on "returns"/"creates" changes until Bob has a chance to chime in.

File: lib/src/list_extensions.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (munificent) [reply lib/src/list_extensions.dart:256]: I would do "Returns" if it were my code.

[the remark this answers is not in the document]
Comment (lrhn) [reply lib/src/list_extensions.dart:256]: Well, I'll agree to disagree then.
To summarize my reasoning: "*Returns*" is never the *best* word to start a DartDoc with.
It's redundant. (All non-`void` methods return something, so it's like adding "when called" to the method. It's a given).
If it's the *most important* thing to say about the method, then either the method should be a getter, or iot's a parameterized-getter-like method, and I'd document it with a noun phrase (so remove the "Returns"). Own up to the fact that the method is described by a noun phrase, don't try to hide it by adding "Returns" in front!
If not, then it shouldn't be the first word of the DartDoc. Say something about what the method actually *does*, not just what it returns. 
Finally, we recommend having a paragraph of the DartDoc starting with "Returns" to document the return value. Having the first (short, one-line) paragraph take over that makes it harder to do a satisfactory description of the returned value, and having two sections starting with "Returns" looks odd and can be confusion.

Tl;dr: Using "Returns noun-phrase" as the primary DartDoc of a method is *technically* following the style guide, but missing the point. It's always *better* to use something else.

But if people want it, they can have it. I'll just keep suggesting improvements!

[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/list_extensions.dart:256]: Using the term "split" could make people think this works analogously to `String.split`, which isn't accurate.

[the remark this answers is not in the document]
Comment (tatumizer) [reply lib/src/list_extensions.dart:256]: carve? (Found by browsing Webster Dictionary)

[the remark this answers is not in the document]
Comment (rakudrama) [reply lib/src/list_extensions.dart:256]: `batched` or `batches`

[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/list_extensions.dart:256]: I think referring to "slices" makes the most sense, since (at least for lists) this is returning views of the sort provided by `slice()`.

File: lib/src/iterable_extensions.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (lrhn) [reply lib/src/iterable_extensions.dart:568]: (Not often encountering a perfectly reasonable language feature isn't a reason not to use it when it doe apply. That just keeps it from being encountered.)

That said, the code here is readable, and makes it clear that there is always at least one element in the yielded list. The rewrite only does that indirectly by having the `add` in a `do`/`while` loop. That's not nearly as visible.

So, OK.

File: lib/src/list_extensions.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (lrhn) [reply lib/src/list_extensions.dart:256]: I agree that `slices` is a an appropriate name. It's a noun phrase, which does mean that it's defined in terms of the values it provides. And that's fair, but the returned value is an iterable, and I'd document *that* in terms of its elements (with a plural). So:
```dart
/// Contiguous [slice]s of [this] with the given [length].
```
Don't say "Returns" (it's trivial). Don't say "Iterable of", that's obvious from the return type. Just describe the elements of that iterable.

I'd still remove the `Returns` and document it with a noun phrase too, for consistency.

File: lib/src/list_extensions.dart [hunk 7]
```diff
@@ -252,6 +252,23 @@ extension ListExtensions<E> on List<E> {
...
+  ///
+  /// Each slice is a view of this list [length] elements long, except for the
+  /// last one which may be shorter if [this] contains too few elements. Each
+  /// slice begins after the last one ends.
+  ///
+  /// As with [slice], these slices are backed by this list, which must not
+  /// change its length while the views are being used.
+  ///
+  /// For example, `[1, 2, 3, 4, 5].slices(2)` returns `[[1, 2], [3, 4], [5]]`.
```
Comment (lrhn) [inline lib/src/list_extensions.dart:265]: I'd say it returns `([1, 2], [3, 4], [5])`, the returned value is not a list, and iterables print themselves using `(...)`.
